### PR TITLE
Added a build to check coding styles

### PR DESCRIPTION
**Pull request by mickaelandrieu**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | We should check ~~our coding style~~ Symfony coding styles on travis, for now only for `src` folder. Why Symfony's ones? Because now we rely on Symfony framework and we have a majority of Symfony developers in the team. More, this close once and for all debates about our "owns" coding styles, as the Symfony ones are really popular (like we did in javascript, relying on airbnb rules)
| Type?         | improvement
| Category?     | CO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | [BOOM-5167](http://forge.prestashop.com/browse/BOOM-5167)
| How to test?  | Build should be green. To fix coding styles, just do ``php bin/php-cs-fixer fix``. You can install the git hook using `composer git-hook-install`.

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/8857)

### CODE REVIEW

**Comment by Quetzacoalt91** [comment]
ping @eternoendless, you had the same idea: https://github.com/PrestaShop/PrestaShop/pull/7979

@mickaelandrieu, you should have made 2 commits for better readability, one for code sniffer addition, the second one for the fixes.

**Comment by mickaelandrieu** [comment]
@Quetzacoalt91 done and rebased to develop branch.

Please don't wait too much before merging this one, this will become very painful to maintain it. The sooner we apply the standards, the better: we are going to freeze the release it's the perfect time to fix every error this pull request can reveal. (Mostly, thanks to the value and type checks rule)

File: .github/contrib/install.php
Comment (Quetzacoalt91) [inline .github/contrib/install.php]: It seems you do the same thing actually. But with a wrong check: "if the file does not exist, get its content".

Perhaps `if (file_exists(__DIR__.'/pre-commit')) {` was enough?

Comment (mickaelandrieu) [reply .github/contrib/install.php]: What do you think about my update?

Comment (Quetzacoalt91) [reply .github/contrib/install.php]: Ah better, now I understand we do not deal with the same file.
But ... If you make a backup of the old file, do you want to copy your new file in all cases? The `else` shouldn't be removed?

**Review by Quetzacoalt91 — DISMISSED** [review]

**Review by Quetzacoalt91 — DISMISSED** [review]

**Comment by LittleBigDev** [comment]
Reviewed 3 of 6 files at r1, 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[.php_cs.dist, line 8 at r2](https://reviewable.io:443/reviews/prestashop/prestashop/8857#-L8H7_xhgtJIfsSWfHRM:-L8H6p1oVuqTXl7-TEfg:bb351kk) ([raw file](https://github.com/prestashop/prestashop/blob/00b5c8533440e5ed0e599dd41ab0d17c69e22d01/.php_cs.dist#L8)):*
> ```PHP
>     ->setRiskyAllowed(false)
>     ->setRules([
>         'array_syntax' => ['syntax' => 'long'],
> ```

Sorry, I'm not ok to force long notation for arrays. We should have the choice (legacy code is forced to long notation, but not new code).
IMHO we should force short notations. But this is not a unanimous point of view, that's why I won't fight for that.
**I'll fight about having the choice tho.**

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8857)*

**Review by LittleBigDev — CHANGES_REQUESTED** [review]
Thank you @mickaelandrieu 
Please see my Reviewable comments.

**Comment by mickaelandrieu** [comment]
> IMHO we should force short notations. But this is not a unanimous point of view, that's why I won't fight for that.

As the validator controls the `src` folder "only" - so only the 1.7 part, our comment is interesting. As we may want to cherry pick a patch in 1.6 from develop, I think "for now" it's ok to keep this rule (devs can type short arrays and then execute the command ^^) and after the end of maintenance of 1.6.1.x it's up to @eternoendless to open a new debate on our coding standards and we can make the described rules updated (to be honest they are FAR less strict than I wanted, but @eternoendless wanted the "soft" way^^).

So, to answer you: I don't think it's in the scope of this pull request to define new standards, but only to fit the current ones.

**Comment by LittleBigDev** [comment]
> only to fit the current ones.

Currently, we can use both notations. The rule would force us to use the long notation.
The code of your PR uses itself the short array notation xD

**Comment by mickaelandrieu** [comment]
> Currently, we can use both notations. The rule would force us to use the long notation.

 I'm not aware of it, and I'm against the use of both notations (it's ugly!). I don't think @eternoendless will validate the use of "both notations" in "src" folder.

> The code of your PR uses itself the short array notation xD

I'll update, it's a copy paste from their docs ^^

**Comment by LittleBigDev** [comment]
Well, we debated a lot (too much) and the consensus was :
- use short notation for everything, except for php <5.4 branches (old supported versions). Same thing for PRs that could be cherry-picked in these old branches.
- When long array syntax is encountered, apply boy scout rule.

I know this is not perfect, but this is the less painful way :/. Changing everything everywhere would be... well... painful.

**Comment by eternoendless** [comment]
Let's not force people into long syntax or short syntax. I don't think it's worth it.

Regarding cherry picks form 1.7 to 1.6, it's extremely rare.

**Comment by mickaelandrieu** [comment]
> Changing everything everywhere would be... well... painful.

I've just executed a command my fingers are well :)

> Let's not force people into long syntax or short syntax. I don't think it's worth it.

I don't think we force everything, except execute a command before commit or merge: you are still free to use long or short arrays when you code, the tool is here to fix everything for you.

Well, I'm little bit tired of this infinite debate that no one is willing to close, we don't have time for it. Let's take it the other way: if we don't have strict rules with a tool to check and fix it, I don't want my pull requests to be stuck because of coding styles and I won't fix "manually" anything related.

If it's ok for everyone, I close.

**Comment by eternoendless** [comment]
We can have strict rules about standardized and generally agreed upon things like PSR-2.

Regarding _those_, I'm in favor of enforcing them and automatically fixing them using an automated tool such as cs-fixer.

For unstandardized things like yoda statements, dot space and array style, we can't automate it because it would force people's code into either one once they run the fixer tool. I don't think it's a big deal so we can leave those alone.

Would that work for you?

**Comment by mickaelandrieu** [comment]
@eternoendless it's not about "me", it's about having a list of rules and rely on it.

If we remove so much rules that even with the fixer our code still looks ugly and different in every part of src folder, that because the fixer validation don't block the merge everyone can still code the way he wants, what is the point with rules and adding a validator?

**Comment by LittleBigDev** [comment]
Code style rules/recommendations are about standardizing code (we all know the benefits of it). Removing too much rules would compromise standardization. The question is : would "dot spaces" and "array notation" compromise standardization if left free ? Because this is what we are talking about. Is it even worth it ?
If yes, we should state something quick because this is becoming ludicrous.
If we really need to debate about it, then here is my personal preference : short array notation and spaces around dots.
Or we can remove rules considered harmless. Your move. But please, state quick, I DO want this cs-fixer merged.

**Comment by mickaelandrieu** [comment]
Thinking about it, I'm not the right person to push this move: only @eternoendless can do it or maybe @Quetzacoalt91.

I close the pull request but don't delete it: feel free to use/enrich/improve it or not.

Thanks for all the reviews :+1:

File: .php_cs.dist
Comment (mickaelandrieu) [inline .php_cs.dist]: from what I see now, this rule is not true anymore

**Comment by mickaelandrieu** [comment]
Hey fellows, this pull request have been refreshed, remove all extra rules and keeping only the ones from Symfony standards.

For long/short arrays debate, let's keep this aside for now and be focused on what we all want: rely on a tool to manage and fix our coding styles and make it part of our validation process.

I hope this time will be the right time to do it, our community is looking at us.

ping @PrestaShop/prestashop-core-dev

**Comment by PierreRambaud** [comment]
Could be cool to add https://github.com/FriendsOfPHP/PHP-CS-Fixer#using-php-cs-fixer-on-ci =)

**Comment by mickaelandrieu** [comment]
@PierreRambaud yes, but it depends if we fix all files now (ie we also accept my second commit) or if we adopt boy scoot rule (we do it everytime we create/update a file).

As it's only a command and doesn't need any brain to do it, I'm in favor of doing it once and then check all the folder every time: wdyt?

**Comment by PierreRambaud** [comment]
Agreeeeeddddd!!!

File: .travis.yml
Comment (eternoendless) [inline .travis.yml:57]: This should be run only on the modified files, and show all failures (not just the first one)

Comment (mickaelandrieu) [reply .travis.yml:57]: @eternoendless depends, see my previous comment :)

Showing all failures doesn't add extra useful information as the only recommended way to fix at least one violation would be to do `php ./vendor/bin/php-cs-fixer fix`.

Comment (eternoendless) [reply .travis.yml:57]: I guess the value is that it makes you learn?

If you decide to keep the `stop-on-violation` rule then it should display a message explaining how to fix the errors so that there's no back-and-forth waiting for the build to pass.

In any case, it will be a lot faster if it only checks modified files.

Comment (PierreRambaud) [reply .travis.yml:57]: Not really faster, maybe 1 or 2 seconds faster for a CI it's nothing :/

Comment (eternoendless) [reply .travis.yml:57]: Let's wait for travis and find out how long it actually takes!

Comment (mickaelandrieu) [reply .travis.yml:57]: it's not up to me to decide ^^ we still need the so famous-called meeting to decide together. Then, I can apply what has been decided.

Also, we can make this step in another build thanks to the jobs feature of Travis: you can look at something I've done in @prestonBot [repository](https://travis-ci.org/PrestaShop/prestonbot/builds/402049713?utm_source=github_status&utm_medium=notification). This way we could  remove ``stop-on-violation`` and I'm almost sure we can ask PrestonBot to ask people to execute the right command to fix CS only if the "CS" jobs fail.

> In any case, it will be a lot faster if it only checks modified files.

I agree, but discussing with @dunglas (member of Symfony core team), I have the feeling that we should apply on all files in case of PHP-CS-Fixer updates. If the issue is only the performance, an extra (and not required) job will do the job (ahah)

Comment (PierreRambaud) [reply .travis.yml:57]: Don't forget that our tests in the CI are really really fast when many open source projects have their tests taking more than one hour with fast testing infrastructure :P

Comment (dunglas) [reply .travis.yml:57]: In API Platform we moved "lint jobs" (PHP CS Fixer and PHPStan) to CircleCI so we can parallelize more: https://github.com/api-platform/core/pull/2004

Also, there is also a free service for OpenSource projects that is blazing fast: https://styleci.io/
It has been created by Laravel maintainers IIRC.

Comment (mickaelandrieu) [reply .travis.yml:57]: Thanks for your feedbacks @dunglas :-)

**Comment by mickaelandrieu** [comment]
Hello fellows, as we have merged first part of this contribution already: I can close this one.

I don't delete the branch *yet*, so you may rely on/reuse some parts.

Thanks, everyone!

**Pull request closed without merging**
